components/Init/src/virtio_vsock.c

```diff
@@ -0,0 +1,243 @@
+{
+    /* Find the CID */
+    for (int i = 0; i < NUM_PORTS; i++) {
+        if (connections[i].cid == cid) {
+            int err = camkes_virtqueue_driver_scatter_send_buffer(connections[i].send_queue, buffer, len);
+            connections[i].send_queue->notify();
+            return;
+        }
+    }
+
+    ZF_LOGE("Invalid CID: %d", cid);
+}
+
+/**
+ * Sets up virtio vsock connections.
+ */
+static void make_virtio_vsock_virtqueues(void)
+{
+    int err;
+    for (int i = 0; i < NUM_PORTS; i++) {
+        virtqueue_driver_t *vq_send;
+        virtqueue_device_t *vq_recv;
+
+        vq_recv = malloc(sizeof(*vq_recv));
+        ZF_LOGF_IF(!vq_recv, "Unable to alloc recv camkes-virtqueue for port: %d", i);
+
+        vq_send = malloc(sizeof(*vq_send));
+        ZF_LOGF_IF(!vq_send, "Unable to alloc send camkes-virtqueue for port: %d", i);
+
+        /* Initialise send virtqueue */
+        err = camkes_virtqueue_driver_init(vq_send, socket_layout[i].send_id);
+        ZF_LOGF_IF(err, "Unable to initialise send camkes-virtqueue for port: %d", i);
+
+        /* Initialise recv virtqueue */
+        err = camkes_virtqueue_device_init(vq_recv, socket_layout[i].recv_id);
+        ZF_LOGF_IF(err, "Unable to initialise recv camkes-virtqueue for port: %d", i);
+
+
+        connections[i].cid = socket_layout[i].cid;
+        connections[i].recv_queue = vq_recv;
+        connections[i].send_queue = vq_send;
+    }
+}
+
+/**
+ * Callback function for the emul layer. Notifies the guest that there is
+ * something in its used ring
+ *
+ * @see virtio_pci_vsock.h for function format
+ *
+ * @param cookie Assumes struct virtio_vsock_cookie
+ */
+static void vsock_inject_irq(void *cookie)
+{
+    virtio_vsock_cookie_t *virtio_cookie = (virtio_vsock_cookie_t *)cookie;
+    if (!virtio_cookie) {
+        ZF_LOGE("NULL virtio cookie given to raw irq handler");
+        return;
+    }
+
+    int err = vm_inject_irq(virtio_cookie->vm->vcpus[BOOT_VCPU], VIRTIO_SCK_IRQ);
+    if (err) {
+        ZF_LOGE("Failed to inject irq");
+        return;
+    }
+}
+
+void make_virtio_vsock(vm_t *vm, vmm_pci_space_t *pci, vmm_io_port_list_t *io_ports)
+{
+    int err;
+
+    struct vsock_passthrough backend;
+    virtio_vsock_cookie_t *vsock_cookie;
+
+    backend.guest_cid = guest_cid;
+    backend.injectIRQ = vsock_inject_irq;
+    backend.forward = tx_forward;
+
+    vsock_cookie = (virtio_vsock_cookie_t *)calloc(1, sizeof(virtio_vsock_cookie_t));
+    if (vsock_cookie == NULL) {
+        ZF_LOGE("Failed to allocated virtio vsock cookie");
+        return;
+    }
+
+    vsock_cookie->vm = vm;
+    backend.vsock_data = (void *) vsock_cookie;
+
+    ioport_range_t virtio_port_range = {0, 0, VIRTIO_IOPORT_SIZE};
+    virtio_vsock = common_make_virtio_vsock(vm, pci, io_ports, virtio_port_range, IOPORT_FREE,
+                                            VIRTIO_SCK_IRQ, VIRTIO_SCK_IRQ, backend);
+
+    make_virtio_vsock_virtqueues();
+    assert(virtio_vsock);
```

Comment:
please move the assert up. There seem not much point in calling `make_virtio_vsock_virtqueues()` when this fails.